File: p146.cpp
#include <bits/stdc++.h>

using namespace std;

struct DoublyLinkedNode {
    int key;
    int val;
    DoublyLinkedNode* next;
    DoublyLinkedNode* prev;
    DoublyLinkedNode(int k, int v) : key(k), val(v), next(NULL), prev(NULL) { }
};

class LRUCache {
public:
    int cap;
    int size;
    unordered_map<int, DoublyLinkedNode*> dictionary;
    DoublyLinkedNode* head;
    DoublyLinkedNode* tail;
    
    LRUCache(int capacity) {
        cap = capacity;
        size = 0;
        head = NULL;
        tail = NULL;
    }
    
    void make_head(int key) {        
        // if it's already the head, do nothing
        if(head == dictionary[key])
            return;
        
        // update tail
        if(tail == dictionary[key]) {
            tail = tail -> prev;
        }
        
        // make the ones beside point to the right place
        DoublyLinkedNode* right = dictionary[key]->next;
        DoublyLinkedNode* left = dictionary[key]->prev;
        if(left != NULL)
            left->next = right;
        if(right != NULL)
            right->prev = left;
        
        // actually make it the head
        dictionary[key]->prev = NULL;        
        head->prev = dictionary[key];
        dictionary[key]->next = head;
        head = dictionary[key];
    }
    
    void set_head(int key, int value) {
        DoublyLinkedNode* front = new DoublyLinkedNode(key, value);
        front->next = head;
        head->prev = front;
        head = front;
    }
    
    int get(int key) {
        // not found
        if(dictionary.find(key) == dictionary.end())
            return -1;
        // get the value
        int value = dictionary[key]->val;
        // do nothing
        if(cap == 1) {
            return value;
        }
        make_head(key);
        return value;        
    }
    
    void put(int key, int value) {
        ++size;
        // set the value
        if(dictionary.find(key) != dictionary.end()) {
            --size;
            dictionary[key]->val = value;
            // we need to move it to head as well
            make_head(key);
        } 
        // over capacity - evict and add
        else if(size > cap) {
            // evict last one
            // case that cap is 1
            if(cap == 1) {
                dictionary.erase(tail->key);
                head = new DoublyLinkedNode(key, value);
                tail = head;            
            } else {
                dictionary.erase(tail->key);
                tail = tail->prev;
                tail->next = NULL;
                delete tail->next;
                set_head(key, value);
            }            
        }
        // still in capacity - just add
        else {
            // add to the head
            if(head == NULL) {
                head = new DoublyLinkedNode(key, value);
                tail = head;
                head->next = NULL;
            } else {
                set_head(key, value);
            }
        }
        dictionary[key] = head;
    }
};

File: p146.py
from collections import defaultdict

class ListNode:
    def __init__(self, k, v):
        self.key = k
        self.val = v
        self.next = None
        self.prev = None

class LRUCache:
    def __init__(self, capacity: int):
        self.head = None
        self.tail = None
        self.lookup = defaultdict(lambda: None)
        self.capacity = capacity
        self.size = 0

    def make_head(self, node: ListNode):
        # it's already the head
        if node is self.head or self.capacity == 1:
            return

        # update tail if necessary
        if node is self.tail:
            self.tail = self.tail.prev

        # update pointers
        if node.prev is not None:
            node.prev.next = node.next
        if node.next is not None:
            node.next.prev = node.prev

        # put it at self.head
        node.prev = None
        node.next = self.head
        self.head.prev = node
        self.head = node
        return

    def get(self, key: int) -> int:
        # try to do a lookup
        ret = self.lookup[key]

        # make head
        if ret is not None:
            self.make_head(ret)
            return ret.val
        else:
            return -1

    def put(self, key: int, value: int) -> None:
        # special case for just updating the node
        if self.get(key) != -1:
            self.head.val = value
            return

        # special case when list is empty
        if self.size == 0:
            self.head = ListNode(key, value)
            self.tail = self.head
            self.lookup[key] = self.head
            self.size += 1
            return

        # special case for capacity == 1 and list is full
        if self.size == 1 and self.capacity == 1:
            del self.lookup[self.head.key]
            self.head.key = key
            self.head.val = value
            self.lookup[key] = self.head
            return

        # capacity >= 2
        # evict if necessary
        if self.size == self.capacity:
            # evict
            tmp = self.tail
            self.tail = self.tail.prev
            del self.lookup[tmp.key]
            del tmp
            self.size -= 1

        # update the head node
        new_head = ListNode(key, value)
        new_head.next = self.head
        self.head.prev = new_head
        self.head = new_head
        self.lookup[key] = self.head
        self.size += 1
